toolchains/extensions.bzl

```diff
@@ -0,0 +1,53 @@
+"""Module extensions for local and remote java repositories"""
+
+load(":local_java_repository.bzl", "local_java_repository")
+load(":remote_java_repository.bzl", "remote_java_repository")
+
+visibility(["//java"])
+
+def _java_repository_impl(mctx):
+    for mod in mctx.modules:
+        for local in mod.tags.local:
+            local_java_repository(
+                local.name,
+                java_home = local.java_home,
+                version = local.version,
+                build_file = local.build_file,
+                build_file_content = local.build_file_content,
+            )
+        for remote in mod.tags.remote:
+            remote_java_repository(
+                remote.name,
+                remote.version,
+                target_compatible_with = remote.target_compatible_with,
+                prefix = remote.prefix,
+                sha256 = remote.sha256,
+                strip_prefix = remote.strip_prefix,
+                urls = remote.urls,
+            )
+
+_local = tag_class(attrs = {
+    "name": attr.string(mandatory = True),
```

Comment:
Module extensions typically don't work well if they accept user-provided names while also collecting tags from transitive deps. There just is no good way to handle collisions.

Since this extension would only wrap simple repos, could `use_repo_rule` be an alternative? If not because the repo rules really are macros, what's blocking them from being rules?

Reply:
Thanks for the heads up. This is just me experimenting for https://github.com/bazelbuild/bazel/issues/26589

If we do go ahead with this, I was imagining only permitting root modules to use this. WDYT?